### PR TITLE
Add multiple key people

### PR DESCRIPTION
**Pull request by irisfaraway**
https://eaflood.atlassian.net/browse/WC-106

Most registrations (excluding sole traders) should be able to add multiple key people. Partnerships must have a minimum of 2.

This doesn't cover removing key people, just adding multiple ones.

### CODE REVIEW

File: app/views/key_people_forms/new.html.erb [hunk 1]
```diff
@@ -1,13 +1,18 @@
 <%= render("shared/back", back_path: back_key_people_forms_path(@key_people_form.reg_identifier)) %>
 
-<div class="text">
+<% if @key_people_form.maximum_key_people == 1 %>
```
Comment (Cruikshanks) [inline app/views/key_people_forms/new.html.erb]: We make reference to `@key_people_form.maximum_key_people == 1` in a number of places in this view, and I think we could better express what we want with a method call.

You also created `can_have_multiple_key_people?` so perhaps `unless can_have_multiple_key_people?` or something new like `only_one_person_allowed?` (said I haven't got my awake head on 😩, sorry)

File: app/forms/key_people_form.rb [hunk 2]
```diff
@@ -48,6 +92,28 @@ def add_key_person
                   person_type: "key")
   end
 
+  def all_key_people
+    # If there's only one key person allowed, just replace existing data
+    return [key_person] if maximum_key_people == 1
+
+    existing_key_people = []
```
[the remark this answers is not in the document]
Comment (irisfaraway) [reply app/forms/key_people_form.rb]: Initially that is what I had! However this resulted in a bug where it actually updated `@transient_registration.keyPeople` immediately when I added to the array. This meant that submitted key people would always be added to the database, even if validations failed. Took me a lot of head-desking to work that one out.

File: app/views/key_people_forms/new.html.erb (hunk 1 above)
Comment (irisfaraway) [reply app/views/key_people_forms/new.html.erb]: That should have been replaced with `can_have_multiple_key_people?`. Will sort it out 👍

File: app/validators/key_people_validator.rb [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
 
   private
 
-  def fields_have_content?(record)
-    fields = [record.first_name, record.last_name, record.dob_day, record.dob_month, record.dob_year]
-    fields.each do |field|
-      return true if field.present? && field.to_s.length.positive?
-    end
-    record.errors.add(:base, :not_enough_key_people)
+  def already_has_enough_key_people?(record)
```
[the remark this answers is not in the document]
Comment (irisfaraway) [reply app/validators/key_people_validator.rb]: Makes sense to me. I agree it is a bit smelly. Will try to tidy it up.

File: app/validators/key_people_validator.rb [hunk 4]
```diff
@@ -1,19 +1,21 @@
 class KeyPeopleValidator < ActiveModel::Validator
   def validate(record)
-    return false unless fields_have_content?(record)
-    validate_first_name(record)
-    validate_last_name(record)
-    DateOfBirthValidator.new.validate(record)
+    if record.fields_have_content?
```
[the remark this answers is not in the document]
Comment (irisfaraway) [reply app/validators/key_people_validator.rb]: No worries. I'll fix up the other issues and then check in for a chat 👍

File: app/views/key_people_forms/new.html.erb (hunk 1 above)
Comment (irisfaraway) [reply app/views/key_people_forms/new.html.erb]: Flipped this around to `can_only_have_one_key_person?` in the end as we were using the inverse of `can_have_multiple_key_people?` more often than not.

**Comment by irisfaraway** [comment]
@Cruikshanks Think I've addressed the issues from the review (all good points, thanks!) – have a look when you've got a minute :)

**Pull request merged**